File: Common/XmlTraverser.vb
Namespace Common

    Public MustInherit Class XmlTraverser

        Public Shared Function Create(ByVal VisitDelegate As Action(Of Xml.XmlNode, VisitAction)) As DelegateXmlTraverser
            Return New DelegateXmlTraverser(VisitDelegate)
        End Function

        Protected MustOverride Sub Visit(ByVal Node As Xml.XmlNode, ByVal VisitAction As VisitAction)

        Public Sub Traverse(ByVal Node As Xml.XmlNode)
            Me._VisitAction.Reset()
            Me.Visit(Node, Me._VisitAction)

            If Me._VisitAction.TraverseChildren Then
                For Each Child As Xml.XmlNode In Node.ChildNodes
                    Me.Traverse(Child)
                Next
            End If
        End Sub

        Private ReadOnly _VisitAction As VisitAction = New VisitAction()

        Public Class VisitAction

            Public Sub New()
                Me.Reset()
            End Sub

            Public Sub Reset()
                Me.TraverseChildren = True
            End Sub

#Region "TraverseChildren Property"
            Private _TraverseChildren As Boolean

            Public Property TraverseChildren As Boolean
                Get
                    Return Me._TraverseChildren
                End Get
                Set(ByVal Value As Boolean)
                    Me._TraverseChildren = Value
                End Set
            End Property
#End Region

        End Class

    End Class

    Public Class DelegateXmlTraverser
        Inherits XmlTraverser

        Public Sub New(ByVal VisitDelegate As Action(Of Xml.XmlNode, VisitAction))
            Me.VisitDelegate = VisitDelegate
        End Sub

        Protected Overrides Sub Visit(Node As Xml.XmlNode, VisitAction As VisitAction)
            Me.VisitDelegate.Invoke(Node, VisitAction)
        End Sub

        Private ReadOnly VisitDelegate As Action(Of Xml.XmlNode, VisitAction)

    End Class

End Namespace

File: Common/XmlTraverser.cs
using System;

namespace Ks.Common
{
    public abstract class XmlTraverser
    {
        public static DelegateXmlTraverser Create(Action<System.Xml.XmlNode, VisitAction> VisitDelegate)
        {
            return new DelegateXmlTraverser(VisitDelegate);
        }

        protected abstract void Visit(System.Xml.XmlNode Node, VisitAction VisitAction);

        public void Traverse(System.Xml.XmlNode Node)
        {
            this._VisitAction.Reset();
            this.Visit(Node, this._VisitAction);

            if (this._VisitAction.TraverseChildren)
            {
                foreach (System.Xml.XmlNode Child in Node.ChildNodes)
                {
                    this.Traverse(Child);
                }
            }
        }

        private readonly VisitAction _VisitAction = new VisitAction();

        public class VisitAction
        {
            public VisitAction()
            {
                this.Reset();
            }

            public void Reset()
            {
                this.TraverseChildren = true;
            }

            public bool TraverseChildren { get; set; }
        }
    }

    public class DelegateXmlTraverser : XmlTraverser
    {
        public DelegateXmlTraverser(Action<System.Xml.XmlNode, VisitAction> VisitDelegate)
        {
            this.VisitDelegate = VisitDelegate;
        }

        protected override void Visit(System.Xml.XmlNode Node, VisitAction VisitAction)
        {
            this.VisitDelegate.Invoke(Node, VisitAction);
        }

        private readonly Action<System.Xml.XmlNode, VisitAction> VisitDelegate;
    }
}
